model/stats/query.go

```diff
@@ -0,0 +1,284 @@
+		catcher.Add(f.StartAt.validateForTests(f.GroupBy))
+	}
+	if len(f.Tests) == 0 && len(f.Tasks) == 0 {
+		catcher.Add(errors.New("Missing Tests or Tasks values"))
+	}
+
+	return catcher.Resolve()
+}
+
+// Validates that the StatsFilter struct is valid for its intended use.
+func (f *StatsFilter) validateForTasks() error {
+	catcher := grip.NewBasicCatcher()
+
+	catcher.Add(f.validateCommon())
+	if f.StartAt != nil {
+		catcher.Add(f.StartAt.validateForTasks(f.GroupBy))
+	}
+	if len(f.Tests) > 0 {
+		catcher.Add(errors.New("Invalid Tests value, should be nil or empty"))
+	}
+	if len(f.Tasks) == 0 {
+		catcher.Add(errors.New("Missing Tasks values"))
+	}
+	if f.GroupBy == GroupByTest {
+		catcher.Add(errors.New("Invalid GroupBy value for a task filter"))
+	}
+
+	return catcher.Resolve()
+
+}
+
+//////////////////////////////
+// Test Statistics Querying //
+//////////////////////////////
+
+// Test execution statistics.
+type TestStats struct {
+	TestFile     string    `bson:"test_file" json:"test_file"`
+	TaskName     string    `bson:"task_name" json:"task_name"`
+	BuildVariant string    `bson:"variant" json:"variant"`
+	Distro       string    `bson:"distro" json:"distro"`
+	Date         time.Time `bson:"date" json:"date"`
+
+	NumPass         int       `bson:"num_pass" json:"num_pass"`
+	NumFail         int       `bson:"num_fail" json:"num_fail"`
+	AvgDurationPass float32   `bson:"avg_duration_pass" json:"avg_duration_pass"`
+	LastUpdate      time.Time `bson:"last_update"`
+}
+
+// Queries the precomputed test statistics using a filter.
+func GetTestStats(filter *StatsFilter) ([]TestStats, error) {
+	err := filter.validateForTests()
+	if err != nil {
+		return nil, err
+	}
+	var stats []TestStats
+	pipeline := testStatsQueryPipeline(filter)
+	err = db.Aggregate(dailyTestStatsCollection, pipeline, &stats)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to aggregate test statistics")
+	}
+	return stats, nil
+}
+
+//////////////////////////////
+// Task Statistics Querying //
+//////////////////////////////
+
+// Task execution statistics.
+type TaskStats struct {
+	TaskName     string    `bson:"task_name" json:"task_name"`
+	BuildVariant string    `bson:"variant" json:"variant"`
+	Distro       string    `bson:"distro" json:"distro"`
+	Date         time.Time `bson:"date" json:"date"`
+
+	NumTotal           int       `bson:"num_total" json:"num_total"`
+	NumSuccess         int       `bson:"num_success" json:"num_success"`
+	NumFailed          int       `bson:"num_failed" json:"num_failed"`
+	NumTimeout         int       `bson:"num_timeout" json:"num_timeout"`
+	NumTestFailed      int       `bson:"num_test_failed" json:"num_test_failed"`
+	NumSystemFailed    int       `bson:"num_system_failed" json:"num_system_failed"`
+	NumSetupFailed     int       `bson:"num_setup_failed" json:"num_setup_failed"`
+	AvgDurationSuccess float32   `bson:"avg_duration_success" json:"avg_duration_success"`
+	LastUpdate         time.Time `bson:"last_update"`
```

Comment:
why do we have json tags on a db model? why does this field not have a json tag?

Reply:
I added them because I thought I would need them for EVG-5412. LastUpdate didn't need to be encoded so it did not have a tag.
I removed them all as they are not needed after all.